openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,602 @@
+        &mut self,
+        leaf_index: LeafIndex,
+        node: &T,
+    ) -> Result<(), ABinaryTreeDiffError> {
+        let node_index = to_node_index(leaf_index);
+        let direct_path = direct_path(node_index, self.size())?;
+        for node_index in &direct_path {
+            self.add_node(*node_index, node.clone())?;
+        }
+        Ok(())
+    }
+
+    /// Sets the nodes in the direct path of the given leaf index to the nodes
+    /// given in the `path`. Returns an error if the given `leaf_index` does not
+    /// correspond to a leaf in the diff or if the given `path` does not have
+    /// the same length as the leaf's direct path.
+    pub(crate) fn set_direct_path(
+        &mut self,
+        leaf_index: LeafIndex,
+        path: Vec<T>,
+    ) -> Result<(), ABinaryTreeDiffError> {
+        let node_index = to_node_index(leaf_index);
+        let direct_path = direct_path(node_index, self.size())?;
+        if path.len() != direct_path.len() {
+            return Err(ABinaryTreeDiffError::PathLengthMismatch);
+        }
+        for (node_index, node) in direct_path.iter().zip(path.into_iter()) {
+            self.add_node(*node_index, node)?;
+        }
+        Ok(())
+    }
+
+    // Functions related to the shared subtree of two given leaves
+    //////////////////////////////////////////////////////////////
+
+    /// Given two leaf indices, returns the position of the shared subtree root
+    /// in the direct path of the first leaf index. Returns an error if both
+    /// leaf indices are identical or if one of the leaf indices does not
+    /// correspond to a leaf in the diff.
+    pub(crate) fn subtree_root_position(
+        &self,
+        leaf_index_1: LeafIndex,
+        leaf_index_2: LeafIndex,
+    ) -> Result<usize, ABinaryTreeDiffError> {
+        // If the given leaf indices are identical, the shared subtree root is
+        // the index itself. Since the index of the leaf itself doesn't appear
+        // in the direct path, we can't return anything meaningful. This check
+        // also ensures that the tree is large enough such that the direct path
+        // is never empty, since if there is a second leaf index (that is within
+        // the bound of the tree), there is a non-leaf root node that is in the
+        // direct path of all leaves.
+        if leaf_index_1 == leaf_index_2 {
+            return Err(ABinaryTreeDiffError::SameLeafError);
+        } else if leaf_index_1 >= self.leaf_count() || leaf_index_2 >= self.leaf_count() {
+            return Err(ABinaryTreeDiffError::OutOfBounds);
+        }
+        let subtree_root_node_index =
+            lowest_common_ancestor(to_node_index(leaf_index_1), to_node_index(leaf_index_2));
+        let leaf_index_1_direct_path = direct_path(to_node_index(leaf_index_1), self.size())?;
+
+        leaf_index_1_direct_path
+            .iter()
+            .position(|&direct_path_node_index| direct_path_node_index == subtree_root_node_index)
```

Comment:
This is a nice additional check but I think it can also get expensive.
How about we only do it for debug builds?

Reply:
No, just nit picking. Let’s leaf it like this.